internal/util/image/layers.go

```diff
@@ -0,0 +1,99 @@
+package image
+
+import (
+	"archive/tar"
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/containerd/containerd/archive"
+	"github.com/containers/image/v5/pkg/blobinfocache/none"
+	"github.com/containers/image/v5/pkg/compression"
+	"github.com/containers/image/v5/types"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	fsutil "github.com/operator-framework/operator-controller/internal/util/fs"
+)
+
+func ForceOwnershipRWX() archive.Filter {
+	return func(h *tar.Header) (bool, error) {
+		h.Uid = os.Getuid()
+		h.Gid = os.Getgid()
+		h.Mode |= 0700
+		return true, nil
+	}
+}
+
+func AllFilters(filters ...archive.Filter) archive.Filter {
+	return func(h *tar.Header) (bool, error) {
+		for _, filter := range filters {
+			keep, err := filter(h)
+			if err != nil {
+				return false, err
+			}
+			if !keep {
+				return false, nil
+			}
+		}
+		return true, nil
+	}
+}
+
+func OnlyPath(p string) archive.Filter {
+	return func(h *tar.Header) (bool, error) {
+		headerPath := path.Clean(strings.TrimPrefix(h.Name, "/"))
+		wantPath := path.Clean(strings.TrimPrefix(p, "/"))
+		relPath, err := filepath.Rel(wantPath, headerPath)
+		if err != nil {
+			return false, fmt.Errorf("error getting relative path: %w", err)
+		}
+		if relPath == ".." || strings.HasPrefix(relPath, "../") {
+			return false, nil
+		}
+		return true, nil
+	}
+}
+
+func ApplyLayersToDisk(ctx context.Context, dest string, img types.Image, imgSrc types.ImageSource, filter archive.Filter) error {
+	var applyOpts []archive.ApplyOpt
+	if filter != nil {
+		applyOpts = append(applyOpts, archive.WithFilter(filter))
+	}
+
+	if err := fsutil.EnsureEmptyDirectory(dest, 0700); err != nil {
+		return fmt.Errorf("error ensuring empty unpack directory: %w", err)
+	}
+	l := log.FromContext(ctx)
+	l.Info("unpacking image", "path", dest)
+	for i, layerInfo := range img.LayerInfos() {
+		if err := func() error {
+			layerReader, _, err := imgSrc.GetBlob(ctx, layerInfo, none.NoCache)
+			if err != nil {
+				return fmt.Errorf("error getting blob for layer[%d]: %w", i, err)
+			}
+			defer layerReader.Close()
+
+			decompressed, _, err := compression.AutoDecompress(layerReader)
+			if err != nil {
+				return fmt.Errorf("auto-decompress failed: %w", err)
+			}
+			defer decompressed.Close()
+
+			if _, err := archive.Apply(ctx, dest, decompressed, applyOpts...); err != nil {
+				return fmt.Errorf("error applying layer[%d]: %w", i, err)
+			}
+			l.Info("applied layer", "layer", i)
+			return nil
+		}(); err != nil {
+			return errors.Join(err, fsutil.DeleteReadOnlyRecursive(dest))
+		}
+	}
+	if err := fsutil.SetReadOnlyRecursive(dest); err != nil {
+		return fmt.Errorf("error making unpack directory read-only: %w", err)
+	}
+	return nil
+}
```

Comment:
to me all of those wrapper functions above are very specific to the logic required in `source` and it might be a good idea to defer placing them in an external dedicated util/<x> package until there is a need to share them, until them it would be better to have them close to where they're being used imho

Reply:
nvm, my bad - I noticed they are actually shared between op-con and catalogd ;)